src/libraries/System.Private.CoreLib/src/System/IO/StreamWriter.cs

```diff
@@ -320,11 +323,55 @@ private void Flush(bool flushStream, bool flushEncoder)
                     (_byteBuffer = new byte[_encoding.GetMaxByteCount(_charBuffer.Length)]);
             }
 
-            int count = _encoder.GetBytes(new ReadOnlySpan<char>(_charBuffer, 0, _charPos), byteBuffer, flushEncoder);
-            _charPos = 0;
-            if (count > 0)
+            if (OperatingSystem.IsIOS() || OperatingSystem.IsTvOS() || OperatingSystem.IsMacCatalyst())
```

Comment:
`StreamWriter` is a public class that can be used for many things. This particular logic changes it's behaviour for few OSes. Could you please create a custom type that derives from `StreamWriter` ,  override `Flush` and move the new logic there? You might need to make some of its private fields `protected internal`.

https://github.com/dotnet/runtime/blob/4d26112c07391f563cb0ca46af1993635399007e/src/libraries/System.Private.CoreLib/src/System/IO/StreamWriter.cs#L277